feat(agent-autoscaling): add initial autoscaling docs (#356)
checkly/docs: platform/private-locations/autoscaling.mdx

<Accordion title="Prerequisites">

- Prometheus V2 metrics are being ingested for your account — the only source for this gauge. See [Exporting Metrics & Data via Prometheus V2](/integrations/observability/prometheus-v2/).
- Checkly Agents are deployed via the [Checkly agent Helm chart](https://github.com/checkly/helm-charts/tree/main/charts/agent) (or an equivalent `Deployment`). See [Kubernetes Deployment](/platform/private-locations/kubernetes-deployment).
- [KEDA](https://keda.sh) is installed in the cluster.

</Accordion>

## The signal

Checkly exposes the `checkly_private_location_check_runs` gauge through the Prometheus V2 exporter. Filtered by `state` and a `private_location_slug_name`, it provides the count of pending and currently-executing check runs in a single Private Location — the signal you drive replica count from.

The relevant `state` values are:

- `queued` — the check run has been scheduled but not yet picked up by an agent.
- `inflight` — the check run is currently being executed by an agent.

<Note>
Short-running checks may be excluded from the gauge because their impact on Private Location capacity is negligible.
</Note>

> 

Reviewer: ```suggestion
Short-running checks may be excluded from the gauge because their impact on Private Location capacity is negligible.
```

How about  mentioning the 1min delay explicitly?

Author: Done!

---
title: 'Autoscaling'
description: 'Autoscale Checkly Agents in a Private Location with KEDA based on queued and in-flight check runs.'
---

> Scale Checkly Agent pods automatically so capacity tracks live load on a Private Location. This page covers the KEDA-based recipe; for static capacity planning, see [Scaling and Redundancy](/platform/private-locations/scaling-redundancy).

Reviewer: ```suggestion
Scale Checkly Agent pods automatically in relation to live load. This page covers the KEDA-based recipe; for static capacity planning, see [Scaling and Redundancy](/platform/private-locations/scaling-redundancy).
```

<Accordion title="Prerequisites">

> - Prometheus V2 metrics are being ingested for your account. See [Exporting Metrics & Data via Prometheus V2](/integrations/observability/prometheus-v2/).

Reviewer: is this optional?

Reviewer: I mean: Can this not be true?

Author: This is required, the metric can only a be accessed via our Prometheus V2 Export endpoint.

Reviewer: Does this mean "Prometheus V2 metrics are being ingested into you Prometheus instance"?

Author: Yes, that's correct.

> - Checkly Agents are deployed as a Kubernetes `Deployment`. See [Kubernetes Deployment](/platform/private-locations/kubernetes-deployment).

Reviewer: Most people will deploy it from the helm chart?

Author: Yes but they may not have Prometheus set-up. Having these two deployment items separate reduces the complexity of using the chart overall. Only if a significant workload is deployed to run on private locations autoscaling becomes necessary.

- [KEDA](https://keda.sh) is installed in the cluster.

</Accordion>

## The signal

> Checkly exposes the `checkly_private_location_check_runs` gauge through the Prometheus V2 exporter. Filtered by `state` and a `private_location_slug_name`, it gives a near-real-time count of pending and currently-executing check runs in a single Private Location — the signal you drive replica count from.

Reviewer: this is the queue length?

Author: Yes queue length + number of messages being processed.

> Checkly exposes the `checkly_private_location_check_runs` gauge through the Prometheus V2 exporter. Filtered by `state` and a `private_location_slug_name`, it gives a near-real-time count of pending and currently-executing check runs in a single Private Location — the signal you drive replica count from.

Reviewer: ```suggestion
Checkly exposes the `checkly_private_location_check_runs` gauge through the Prometheus V2 exporter. Filtered by `state` and a `private_location_slug_name`, it provides the count of pending and currently-executing check runs in a single Private Location — the signal you drive replica count from.
```

No promises

The relevant `state` values are:

- `queued` — the check run has been scheduled but not yet picked up by an agent.
- `inflight` — the check run is currently being executed by an agent.

> Summing both gives total live load and is what you want to scale on.

Reviewer: This bit might confuse people, lets just go for sum(metric) instead to show directly what we want the customer to do

<Note>
Short-running checks may be excluded from the gauge because their impact on Private Location capacity is negligible.
</Note>

> ## KEDA `ScaledObject`

Reviewer: Also make this object part of the helm chart?

Author: We are not including it in the helm chart for now atleast. It requires Prometheus V2 import to be used and some users may opt-in to use the helm chart but not Prometheus metrics.

```yaml
apiVersion: keda.sh/v1alpha1
kind: ScaledObject
metadata:
  name: checkly-agent-autoscaler
spec:
  scaleTargetRef:
    kind: Deployment                            # Optional. Default: Deployment.
    namespace: <namespace_for_agent_deployment> # Optional.
    name: <agent_deployment_name>               # Required.
  minReplicaCount: 2                            # Optional. Default: 0.
  maxReplicaCount: 10                           # Optional. Default: 100.

>   pollingInterval: 30                           # Optional. Default: 30 (seconds).

Reviewer: remove defaults?

>   cooldownPeriod: 300                           # Optional. Default: 300. Scale-to-zero only; see below.

Reviewer: remove. cooldown period is only relevant when there is minReplas = 0 https://keda.sh/docs/2.19/reference/scaledobject-spec/#cooldownperiod

  triggers:
    - type: prometheus
      metadata:
        serverAddress: http://prometheus-k8s.monitoring.svc.cluster.local:9090
        metricName: checkly_private_location_check_runs

>         threshold: "2"

Reviewer: this needs to be 1 or equal to the agent parallelism

Reviewer: in the default config one agent can only process 1 check, this will cause long scheduling delay when there are large amounts of long running checks

        query: sum(checkly_private_location_check_runs{state=~"queued|inflight", private_location_slug_name="<private_location_slug>"})
```

The query is scoped to a single Private Location by `private_location_slug_name`, so create one `ScaledObject` per Private Location.

For a Prometheus instance outside the cluster, add an [`authenticationRef`](https://keda.sh/docs/latest/scalers/prometheus/) pointing at a `TriggerAuthentication` resource with the appropriate credentials.

## How desired replica count is computed

> KEDA publishes the PromQL result as an external metric and the HPA it creates applies:

Reviewer: confusing

```
desiredReplicas = ceil( PromQL_result / threshold )
```

clamped to `[minReplicaCount, maxReplicaCount]`. `threshold` is a per-pod target value, not a trigger boundary.

> With `threshold: "2"`, `minReplicaCount: 2`, `maxReplicaCount: 10`:

Reviewer: i'D really recommend 1 threshold

> | `sum(...)` result | `ceil(result / 2)` | After clamp | Resulting pods |

Reviewer: maybe use real human words instead of mathy stuff here

ceil, clamp, this is all very abstract

|---|---|---|---|
| 0  | 0  | 2  | 2 (idle floor) |
| 1  | 1  | 2  | 2 |
| 4  | 2  | 2  | 2 |
| 5  | 3  | 3  | 3 |
| 11 | 6  | 6  | 6 |
| 25 | 13 | 10 | 10 (capped) |

## Tuning the bounds

> - **`threshold`** — lower values produce more pods per unit of work. A reasonable starting point is the per-agent `JOB_CONCURRENCY` configured on the Deployment; halve it to add headroom.

Reviewer: This. Make it a comment in the spec

> - **`minReplicaCount`** — the redundancy floor. Keep at `2` or higher so a single agent failure doesn't take the Private Location offline. See [Scaling and Redundancy](/platform/private-locations/scaling-redundancy).

Reviewer: Add: If 0, cooldown value is important

- **`maxReplicaCount`** — must exceed `ceil(peak_load / threshold)`. If the cap is too low, queued check runs accumulate above it and are dropped after the 6-minute queue TTL.

## Scale-down behavior and `cooldownPeriod`

`cooldownPeriod` governs scale-to-zero only — the wait between the trigger going inactive and KEDA scaling the workload to `0`. With `minReplicaCount: 2`, scale-to-zero never happens, so this field is dormant.

Scale-in between `maxReplicaCount` and `minReplicaCount` is controlled by the HPA's own scale-down stabilization window (default 300 seconds). To dampen scale-in more aggressively, override the HPA behavior on the `ScaledObject`:

```yaml

>   advanced:

Reviewer: i'd recommend adding this to the example spec, the scaleUp and scaleDown parameters are very hard to get right and giving some sane default values would provide quite some value to customers.

Author: I've added our own internal values with additional stabilisation window. 👍

    horizontalPodAutoscalerConfig:
      behavior:
        scaleDown:

>           stabilizationWindowSeconds: 600

Reviewer: Far too high. this will only scale every 10 min

          policies:
            - type: Pods
              value: 1
              periodSeconds: 60
```

## Graceful termination

In-flight checks on a terminating pod are rerun on another agent after a 300-second timeout. Set `terminationGracePeriodSeconds` above this on the agent pod spec so an evicted pod has room to drain before `SIGKILL`:

```yaml
spec:
  template:
    spec:

>       terminationGracePeriodSeconds: 330

Reviewer: this should be set on the helm chart actually. Also this needs to be min as high as the longest check (e.g. PWS up to 30 min)

Author: Done!
